src/Domain/CoordinateMaps/Protocols.hpp

```diff
@@ -0,0 +1,117 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include <array>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+#include "DataStructures/DataVector.hpp"
+#pragma once
+
+/// \ref protocols related to coordinate maps
+namespace domain::CoordinateMaps::protocols {
+
+/*!
+ * \ingroup ProtocolsGroup
+ * \brief Defines the interface for transition functions used by the shape map.
+ *
+ * \details This protocol defines the interface that a class must conform to so
+ * that it may be used as a transition function by the shape map. Different
+ * domains require the shape map to fall off towards the boundary in different
+ * ways. This behavior is controlled by the transition function. It is also
+ * needed to find the inverse of the shape map. Since the shape map preserves
+ * angles, the problem of finding its inverse reduces to the 1-dimensional
+ * problem of finding the original radius from the mapped radius. The mapped
+ * radius \f$\tilde{r}\f$ is related to the original \f$r\f$ radius by:
+ * \f{equation}{
+ * \tilde{r} = r (1 - f(r,\theta,\phi) \sum_{lm} \lambda_{lm}(t)Y_{lm}(\theta,
+ * \phi))
+ * \f}
+ * where everything but \f$\tilde{r}\f$ and \f$r\f$ can be treated as constant.
+ * This equation is inverted in `original_radius_over_radius` and also divided
+ * by the mapped radius to simplify calculations in the shape map.
+ *
+ * The transition function must also be able to compute its gradient and the
+ * value of the function divided by the radius. Care must be taken that this
+ * does not divide by zero.
+ *
+ * All member functions are templated on a parameter T which can be both a
+ * `double` or a `DataVector`. The `ConformingType` needs to have these member
+ * functions:
+ * - `T operator()(std::array<T> source_coordinates)`: Evaluate the transition
+ * function at the Cartesian coordinates.
+ * - `T original_radius_over_radius(std::array<T> target_coordinates)`:
+ * Evaluates the original radius from the mapped radius by inverting the shape
+ * map. It also divides by the mapped radius to simplify calculations in the
+ * shape map.
+ * - `std::array<T> gradient(std::array<T> source_coordinates)`: Evaluate the
+ * gradient of the transition function at the Cartesian coordinates.
+ * - `T map_over_radius(std::array<T> source_coordinates)`: Evaluate the
+ * transition function at the Cartesian coordinates divided by the radius. Care
+ * must be taken not to divide by zero.
+ * - `bool operator==(const ConformingType& lhs, const ConformingType& rhs)`
+ */
+struct TransitionFunc {
```

Comment:
Do you think it might be more appropriate to make this a base class, instead of a protocol? I see the issue with function templates and inheritance, but this class looks a lot like a base class :) On the other hand, a compile-time interface like "the type must have these type aliases" cannot be represented well as a base class, so it's a job for a protocol (e.g. things we put in the Metavariables). It's quite clear that e.g. an `EquationOfState` should be a base class (we do polymorphism, factory creation etc with it) and an evolution or elliptic system should conform to a protocol (we select it in the metavars, it's entirely compile-time). I haven't yet figured out where exactly to draw the line between the two, so perhaps you have some insight here?

Reply:
I initially wanted to make it a base class but dont see how. All the member functions are templated on type DataVector/double whereas the class itself is not. One could maybe template the abstract base transition function on the type and then have the shape map hold both types as a member function. Otherwise you would have to do something fancy (read: ugly) with type erasure.

On a related note, this is yet another instance where the fact that the map member functions are templated for double/DataVector becomes very annoying. I find that in the implementation I spend a very large amount of time trying to find type traits that (1) work for doubles and DataVectors and (2) do not make additional allocations in case of DataVectors. This is sometimes possible only at a large readability cost. From what I understood, the double instances are essentially only used for testing and think that it would save a lot of programming time to get rid of this template and test with DataVectors only (also it would improve readability), perhaps providing some convenience functions to pack/unpack a double into a DataVector. Or will this template be used for something else in the future?

Reply:
Yes, the base class can't define the function templates. It can only document them, but that may be enough. So the alternative here is to make `TransitionFunc` (I suggest to name `TransitionFunction`, no reason to abbreviate) a base class that defines as much as it can, and documents the rest. If we remove some `double`/`DataVector` templates then more functions can be defined by the base class. I think this is the difference to the main use-cases for protocols: They define _compile-time_ interfaces that can never be base classes, like the type aliases in the Metavariables. Again, I haven't figured out exactly where to draw the line, but this seems like a useful distinction. WDYT?

Re coord-map templates: Yes, let's discuss on Slack or in the spectre call.

Reply:
I am not sure that would work. So, I create an abstract base class for TransitionFunction with some documentation but (at the moment) not any actual virtual methods. The actual implementations then derive from this class and implement their own methods which are non-overriding. The Shape map then holds a pointer to the abstract base class. But how can I call the actual methods now from the pointer to an abstract base class? Would I need to loop over all derived classes with a `dynamic_cast` until the cast is successful?

Reply:
The shape map can still be templated on the derived class if that makes more sense in this case than the looping-over-derived-classes (the "fake-virtual" mechanism).

Reply:
ah I see, you have to explain what this fake-virtual actually does to me at some point^^. Yes that should work but it is pretty much exactly the same thing as a protocol except that the protocol allows us to also check that it is conforming with a (static) unit test. I do not really see any reason to change it now unless we get rid of the templates for the methods and use proper virtual methods, no?

Reply:
yep I agree, it's pretty much the same at this point. The difference is that the `TransitionFunction` has a chance to become an actual base class, whereas other protocol use-cases don't. I don't have a strong opinion which way to go in this PR.

Reply:
If there are only two valid types for a template parameter, you can change it to an overload, which can be virtual.  The implementations can then be just calls to non-virtual templated functions, to avoid most of the code duplication.  The `CoordinateMap` class does this.

Reply:
I see, that would work. Do we want to change it at this point? It would make writing new `TransitionFunctions` moderately more tedious especially to someone unfamiliar with the design pattern. Though having overridden member functions is always a nicer alternative since the signature checking is done automatically.

Reply:
@nilsleiffischer any opinions?

Reply:
I have changed it to an abstract base class now. Your third idea seems very rigid to me, I am not sure if it would even be possible once we have more than one transition function. The shape map would have to hold all their data but then there would be no way of knowing which data to pass into the transition function